calyx/src/ir/structure.rs

```diff
@@ -15,21 +15,23 @@ pub enum Direction {
 }
 
 /// Ports can come from Cells or Groups
-#[derive(Debug)]
+#[derive(Debug, Clone)]
 pub enum PortParent {
     Cell(WRC<Cell>),
     Group(WRC<Group>),
 }
 
 /// Represents a port on a cell.
-#[derive(Debug)]
+#[derive(Debug, Clone)]
 pub struct Port {
     /// Name of the port
     pub name: Id,
     /// Width of the port
     pub width: u64,
     /// Direction of the port
     pub direction: Direction,
+    /// Name of parent
+    pub parent_name: Id,
```

Comment:
I remain unconvinced that this is needed. Whenever you're ready, do mention where it is needed in the writeup for the PR.

Reply:
Otherwise there is no way to mutate assignments based on whether it uses a hole from a particular group. This is because ports themselves are not unique and if you are mutating the group, then you can't look up the parent of a hole, because the group is already borrowed. You can't do this in two steps either, because there is no way to relate the immutable information gathering pass to the mutable assignment changing pass because again, ports are not unique. 

If you are really opposed to adding `parent_name`, then we can make ports unique in some other way. But that seems like a more complicated solution than is called for

Reply:
Can you point out the place in the inliner where this logic applies?

My problem with exposing the `parent_name` is that we're back to square one: what is the "canonical" way of getting a parent? Do I use the parent pointer or do I get the parent name and call one of the `find_` methods on the components.

The principles of the IR is to have one way to do things and to not duplicate information. This violates both.